File: 560-subarray-sum-equals-k/subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> mp;
        mp[0] = 1;
        int sum = 0, n = nums.size();
        // 0 1 2
        // 1 1 1
        // 1:0 2:1 3:2
        int c = 0;
        for(int i=0; i<n; i++){
            sum += nums[i];
            int y = sum - k;
            if(mp.find(y) != mp.end()){
                c += mp[y];
            }
            mp[sum] += 1;
            // x + y = k
            
        }
        return c;
    }
};

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int n = nums.length;
        HashMap<Integer, Integer> mp = new HashMap<>();
        int currSum = 0;
        int count = 0;
        
        for(int i = 0; i < n; i++){
            currSum += nums[i];
            if(currSum == k) count++;
            if(mp.containsKey(currSum - k))
                count += mp.get(currSum - k);
            mp.put(currSum, mp.getOrDefault(currSum, 0) + 1);
        }
        return count;
    }
}